stubs/urllib3/urllib3/util/retry.pyi

```diff
@@ -9,42 +13,71 @@ ProtocolError = exceptions.ProtocolError
 ReadTimeoutError = exceptions.ReadTimeoutError
 ResponseError = exceptions.ResponseError
 
-log: Any
+log: logging.Logger
+
+class RequestHistory(NamedTuple):
+    method: str | None
+    url: str | None
+    error: Exception | None
+    status: int | None
+    redirect_location: str | None
 
 class Retry:
-    DEFAULT_METHOD_WHITELIST: Any
-    BACKOFF_MAX: Any
-    total: Any
-    connect: Any
-    read: Any
-    redirect: Any
-    status_forcelist: Any
-    method_whitelist: Any
-    backoff_factor: Any
-    raise_on_redirect: Any
+    DEFAULT_ALLOWED_METHODS: ClassVar[frozenset[str]]
+    RETRY_AFTER_STATUS_CODES: ClassVar[frozenset[int]]
+    DEFAULT_REMOVE_HEADERS_ON_REDIRECT: ClassVar[frozenset[str]]
+    DEFAULT_BACKOFF_MAX: ClassVar[int]
+
+    total: bool | int | None
+    connect: int | None
+    read: int | None
+    redirect: Literal[True] | int | None
+    status: int | None
+    other: int | None
+    allowed_methods: Collection[str] | None
+    status_forcelist: Collection[int]
+    backoff_factor: float
+    raise_on_redirect: bool
+    raise_on_status: bool
+    history: tuple[RequestHistory, ...]
+    respect_retry_after_header: bool
+    remove_headers_on_redirect: frozenset[str]
     def __init__(
         self,
-        total=...,
-        connect=...,
-        read=...,
-        redirect=...,
-        status=...,
-        other=...,
-        allowed_methods=...,
-        status_forcelist=...,
-        backoff_factor=...,
-        raise_on_redirect=...,
-        raise_on_status=...,
-        history=...,
-        respect_retry_after_header=...,
-        remove_headers_on_redirect=...,
-        method_whitelist=...,
+        total: bool | int | None = ...,
+        connect: int | None = ...,
+        read: int | None = ...,
+        redirect: bool | int | None = ...,
+        status: int | None = ...,
+        other: int | None = ...,
+        allowed_methods: Collection[str] | None = ...,
+        status_forcelist: Collection[int] | None = ...,
+        backoff_factor: float = ...,
+        raise_on_redirect: bool = ...,
+        raise_on_status: bool = ...,
+        history: tuple[RequestHistory, ...] | None = ...,
+        respect_retry_after_header: bool = ...,
+        remove_headers_on_redirect: Collection[str] = ...,
+        method_whitelist: Collection[str] | None = ...,
     ) -> None: ...
-    def new(self, **kw): ...
+    def new(self, **kw: Any) -> Retry: ...
```

Comment:
[At runtime](https://github.com/urllib3/urllib3/blob/33afb5c96c435ab7118b5f41e26f96c290fef55b/src/urllib3/util/retry.py#L263) this will return an instance of a subclass if it's being called on a subclass of `Retry`, so we should use a `TypeVar` to annotate the return-type instead of hardcoding it. (This is different to how they've annotated their code on `main`, but I think their annotation is wrong.) typeshed's established style is to use the `_typeshed.Self` `TypeVar` for these situations (you'll need to do `from _typeshed import Self` at the top of the file).

```suggestion
    def new(self: Self, **kw: Any) -> Self: ...
```